moveit_experimental/jog_arm/include/jog_arm/jog_arm_server.h

```diff
@@ -0,0 +1,284 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+// Server node for arm jogging with MoveIt.
+
+#ifndef JOG_ARM_SERVER_H
+#define JOG_ARM_SERVER_H
+
+#include <Eigen/Eigenvalues>
+#include <moveit_experimental/JogJoint.h>
+#include <moveit/move_group_interface/move_group_interface.h>
+#include <moveit/planning_scene/planning_scene.h>
+#include <moveit/planning_scene_interface/planning_scene_interface.h>
+#include <moveit/robot_model_loader/robot_model_loader.h>
+#include <moveit/robot_state/robot_state.h>
+#include <rosparam_shortcuts/rosparam_shortcuts.h>
+#include <sensor_msgs/JointState.h>
+#include <sensor_msgs/Joy.h>
+#include <std_msgs/Bool.h>
+#include <std_msgs/Float64MultiArray.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2_ros/transform_listener.h>
+#include <trajectory_msgs/JointTrajectory.h>
+
+namespace jog_arm
+{
+// Variables to share between threads, and their mutexes
+struct jog_arm_shared
+{
+  geometry_msgs::TwistStamped command_deltas;
+  pthread_mutex_t command_deltas_mutex;
+
+  moveit_experimental::JogJoint joint_command_deltas;
+  pthread_mutex_t joint_command_deltas_mutex;
+
+  sensor_msgs::JointState joints;
+  pthread_mutex_t joints_mutex;
+
+  double collision_velocity_scale = 1;
+  pthread_mutex_t collision_velocity_scale_mutex;
+
+  // Indicates that an incoming Cartesian command is all zero velocities
+  bool zero_cartesian_cmd_flag = true;
+  pthread_mutex_t zero_cartesian_cmd_flag_mutex;
+
+  // Indicates that an incoming joint angle command is all zero velocities
+  bool zero_joint_cmd_flag = true;
+  pthread_mutex_t zero_joint_cmd_flag_mutex;
+
+  // Indicates that we have not received a new command in some time
+  bool command_is_stale = false;
+  pthread_mutex_t command_is_stale_mutex;
+
+  // The new trajectory which is calculated
+  trajectory_msgs::JointTrajectory new_traj;
+  pthread_mutex_t new_traj_mutex;
+
+  // Timestamp of incoming commands
+  ros::Time incoming_cmd_stamp = ros::Time(0.);
+  pthread_mutex_t incoming_cmd_stamp_mutex;
+
+  bool ok_to_publish = false;
+  pthread_mutex_t ok_to_publish_mutex;
+};
+
+// ROS params to be read
+struct JogArmParameters
+{
+  std::string move_group_name, joint_topic, cartesian_command_in_topic, command_frame, command_out_topic,
+      planning_frame, warning_topic, joint_command_in_topic, command_in_type, command_out_type;
+  double linear_scale, rotational_scale, joint_scale, lower_singularity_threshold, hard_stop_singularity_threshold,
+      lower_collision_proximity_threshold, hard_stop_collision_proximity_threshold, low_pass_filter_coeff,
+      publish_period, publish_delay, incoming_command_timeout, joint_limit_margin, collision_check_rate;
+  bool gazebo, collision_check, publish_joint_positions, publish_joint_velocities, publish_joint_accelerations;
```

Comment:
these variable names could be better - what does ``gazebo`` do? at least add comments... but typically you should first have a verb e.g. ``use_gazebo``